### ISSUE TITLE
Race condition, use-after-free in win32 event system

### ISSUE THREAD

**Issue by Trzik** [c-ares/c-ares#798]
Since version 1.28.1 we started using event loop provided by c-ares on Windows. However there seems to be some kind of race condition or use-after-free. Happens very rarely (about once per 2.5k executions I believe, very hard to reproduce). Version 1.29.0 is affected as well, did not try more recent ones.

What we are doing (in a single thread):
* Opening a channel via `ares_init_options` with `ARES_EVSYS_DEFAULT`, options mask is set to `ARES_OPT_EVENT_THREAD`.
* Sending over a hundred queries (117 to be exact) on the same channel via multiple calls to `ares_query_dnsrec`. If it matters, all queries have type `ARES_REC_TYPE_A` and class `ARES_CLASS_IN`.
* We then call `ares_queue_wait_empty` with 1s timeout in a loop where we also check an external terminating flag and a local timeout for the whole operation.
* If `ares_queue_wait_empty` returns `ARES_ETIMEOUT` and we need to terminate the loop (either we ran out of time or the termination flag is set), we call `ares_cancel`. If `ares_queue_wait_empty` returns `ARES_SUCCESS`, we just terminate the loop without calling `ares_cancel`.
* We destroy the channel with `ares_destroy`.

Full stack trace from the event thread (from v1.28.1) at the point of crash where it accessed invalid memory:
```
ares_evsys_win32_afd_enqueue(ares_event * event, ares_event_flags_t flags) Line 334
ares_evsys_win32_wait(ares_event_thread * e, unsigned long timeout_ms) Line 581
ares_event_thread(void * arg) Line 269
ares__thread_func(void * lpParameter) Line 151
```
The `event` pointer looks like a deallocated and reused memory (it points to a valid string but definitely not to a valid event structure).

This is stack trace from our thread at the same point of time:
```
ares__thread_join(ares__thread * thread, void * * rv) Line 189
ares_event_thread_destroy_int(ares_event_thread * e) Line 298
ares_destroy(ares_channeldata * channel) Line 82
```

It looks like there is some synchronization issue when event thread is doing its thing (but there shouldn't really be any queries alive at that point) while we are trying to destroy it.

**Comment by bradh352** [comment]
A couple of questions, are you only testing windows, or are you also testing linux and linux is fine?

Next, if you can try the current 'main', that would be helpful.  I actually worked through some thread-related issues yesterday that are not in any release.

**Comment by bradh352** [comment]
have you been able to reproduce this yourself, or only from user reports?

**Comment by Trzik** [comment]
No, this happens randomly and as I said, very rarely. We have no way to reliably reproduce this, it did not even occur in our nightly tests. We only have a handful of reports since we switched to the event system shortly after version 1.28.1 was released.

**Comment by bradh352** [comment]
I think maybe there is still an issue as I just saw a test failure on something else I'm working on for Windows x86 with MSVC in a Cancel code path:
https://ci.appveyor.com/project/c-ares/c-ares/builds/50117719/job/uabbb941ntb6aa1v

What's interesting is we run this with ASAN on clang64 and nothing nasty gets reported.  

Can you tell me if your crashes are specific to 32bit vs 64bit?

**Comment by bradh352** [comment]
Also, what compiler is being used when you see the crashes?

**Comment by bradh352** [comment]
initial look at debugging this crash I observed appears to be an issue in the test system itself, so don't think this is related to your crash.  I'll know once I finish the test system correction.

**Comment by bradh352** [comment]
c-ares 1.32.0 is out with the aforementioned fixes.  hopefully this version proves stable with EventThread on Windows for you.

**Comment by Trzik** [comment]
Hi, sorry for the delay. To answer your questions and provide more information:
The crash seems to be specific to Windows 7 (build 7601), both 32b and 64b versions are affected. The issue was not observed on any other OS. Windows 10 and 11 seem to be fine (so far?).
On Windows we compile with MSVC, usually using the latest public release.
We tried the recent changes you provided (build from commit f7995109014c85ae92e012e2b07e26f6a35897f4) and sadly there is one crash report with the same issue, so it seems that it did not work 😢.

**Comment by bradh352** [comment]
Do you happen to have another backtrace I can evaluate?  Just to see if they're the same?

So all crashes are on Windows 7 though? Hrm, I don't think I have access to that to test unless microsoft provides a download for that still somewhere.

I did find a windows 32bit issue just yesterday in 8f265c9d, but that wouldn't affect 64bit.

**Comment by bradh352** [comment]
Actually, just realized we always set _WIN32_WINNT to 0602 which is Windows 8 when compiling c-ares. Wonder if this could be related somehow, as the size of some structures could be different than what Windows 7 provides if the program still runs at all on Windows 7.

**Comment by bradh352** [comment]
I did add a test case for calling ares_canel() with a lot of pending requests. I haven't seen it fail, however. #803 has that test case.

There are a few other changes in the v1.32 branch that are not yet released. https://github.com/c-ares/c-ares/commits/v1.32/

**Comment by Trzik** [comment]
Another report came in today. Windows 7, 32b. Note that this is still 1.29 branch build from the commit mentioned somewhere above.

```
ares_evsys_win32_afd_enqueue(ares_event * event, ares_event_flags_t flags) Line 334
ares_evsys_win32_wait(ares_event_thread * e, unsigned long timeout_ms) Line 581
ares_event_thread(void * arg) Line 269
ares__thread_func(void * lpParameter) Line 151
```

From what I can see in the dump, the variable `e` passed to `ares_evsys_win32_wait` is perfectly valid. It contains reasonable data. The completion key obtained from `GetQueuedCompletionStatusEx` seems to be valid as well and points to a valid event data structure. However the event pointer in the very first entry (`i == 0`) points to a complete garbage (see image below). The pointer value itself looks reasonable enough so it really seems that it points to something that was already freed and reused. It's hard to imagine why an issue like this would affect Windows 7 exclusively though.

![dump](https://github.com/c-ares/c-ares/assets/121808562/65c78ae1-6538-4163-8a77-413053b2a954)

**Comment by bradh352** [comment]
well, i'm hoping a later c-ares version will magically fix it.  it is suspicious that your only reports are on windows 7 though.

since 1.29 we did start running windows through the clang static analyzer, and also run it through ASAN and UBSAN.  We also added tests specific to cancelling a bunch of outstanding queries followed by destroy like you stated is your case.

**Comment by bradh352** [comment]
I've been racking my brain over how there may be a use after free in this case.  The only thing I can come up with is if ares_evsys_win32_afd_cancel() returns false, which means there is no existing wait, and cleans up the data structure for the connection (which also closes the socket that was passed to CreateIoCompletionPort()).  THEN, the next call to GetQueuedCompletionStatusEx() returns that pointer that was freed, which according to the docs shouldn't happen as closing the enqueued socket will remove it from the IOCP port.

Of course we're calling into AFD directly which isn't exactly documented.  Maybe the queued completion is already sitting there waiting to be picked up and it doesn't evict an existing result even though we effectively removed it from the IOCP port ... I really don't know.

I need to think about it a bit more.  Its possible some stuff can be reworked with the assumption that we will always receive an IOCP event and to clean up *after* that, rather than before.  Likely I'd need more error handling around ares_evsys_win32_afd_enqueue() as well to clean up if that fails as if it fails we'd never get an event triggered.

**Comment by bradh352** [comment]
So i decided to try to switch up the entire methodology used, have a PR I'm working on in #811.  I still need to do some fine tuning and cleanups.  But I've been running the test suite in an endless loop on Win11 ARM and haven't hit anything odd (I need to boot up a system to run on x86/x64).  

The prior way it was working I found some memory leaks during shutdown and when I tried to correct those I started getting some odd crashes that were hard to resolve.

I will say debugging anything on windows is a real pain.  No LSAN, no Valgrind ... and lldb has some odd issues... ugh.
